main.c

```diff
@@ -182,36 +190,103 @@ static const struct wl_pointer_listener pointer_listener = {
 	.axis = noop,
 };
 
+static void keyboard_handle_keymap(void *data, struct wl_keyboard *wl_keyboard,
+		uint32_t format, int32_t fd, uint32_t size) {
+	struct slurp_seat *seat = data;
+	switch (format) {
+	case WL_KEYBOARD_KEYMAP_FORMAT_NO_KEYMAP:
+		fprintf(stderr, "no keymap\n");
+		seat->xkb_keymap = xkb_keymap_new_from_names(seat->state->xkb_context, NULL, XKB_KEYMAP_COMPILE_NO_FLAGS);
+		break;
+	case WL_KEYBOARD_KEYMAP_FORMAT_XKB_V1:;
+		fprintf(stderr, "libxkbcommon compatible\n");
+		clock_t begin, end;
+		if ((begin = clock()) == -1) {
+			fprintf(stderr, "clock\n");
+			exit(EXIT_FAILURE);
+		}
+
+		// Methods:
+		// 1. convert file descriptor to FILE *
+		// 2. mmap
+		// 3. Read to string (use from_buffer)
+
+#if 0
+		FILE *file = fdopen(fd, "r");
+		seat->xkb_keymap = xkb_keymap_new_from_file(seat->state->xkb_context,
+				file, XKB_KEYMAP_FORMAT_TEXT_V1,
+				XKB_KEYMAP_COMPILE_NO_FLAGS);
+		fclose(file);
+#endif
+
+#if 1
+		void *buffer;
+#if 0
+		if ((buffer = mmap(NULL, size - 1, PROT_READ, MAP_SHARED, fd, 0)) == MAP_FAILED) {
+			err(1, NULL);
+		}
+#endif
+
+#if 1
+		buffer = alloca(size - 1);
+		ssize_t readed = read(fd, buffer, size - 1);
+		assert(readed == size - 1);
+#endif
+		seat->xkb_keymap =
+			xkb_keymap_new_from_buffer(seat->state->xkb_context,
+					buffer, size - 1,
+					XKB_KEYMAP_FORMAT_TEXT_V1,
+					XKB_KEYMAP_COMPILE_NO_FLAGS);
+		close(fd);
+#endif
+
+		if ((end = clock()) == -1) {
+			fprintf(stderr, "clock - end\n");
+			exit(EXIT_FAILURE);
+		}
+
+		fprintf(stderr, "time to parse: %ld\n", end - begin);
+		break;
+	}
+	seat->xkb_state = xkb_state_new(seat->xkb_keymap);
+}
+
 static void keyboard_handle_key(void *data, struct wl_keyboard *wl_keyboard,
-		uint32_t serial, uint32_t time, uint32_t key, uint32_t key_state) {
+		const uint32_t serial, const uint32_t time, const uint32_t key,
+		const uint32_t key_state) {
 	struct slurp_seat *seat = data;
 	struct slurp_state *state = seat->state;
-	if (key_state == WL_KEYBOARD_KEY_STATE_PRESSED) {
-		switch (key) {
-		case KEY_ESC:
+	const xkb_keysym_t keysym = xkb_state_key_get_one_sym(seat->xkb_state, key + 8);
+
+	switch (key_state) {
+	case WL_KEYBOARD_KEY_STATE_PRESSED:
+		switch (keysym) {
+		case XKB_KEY_Escape:
 			seat->has_selection = false;
 			state->edit_anchor = false;
 			state->running = false;
 			break;
 
-		case KEY_SPACE:
+		case XKB_KEY_space:
 			if (!seat->has_selection) {
 				break;
 			}
 			state->edit_anchor = true;
 			break;
 		}
-	}
+		break;
 
-	if (key_state == WL_KEYBOARD_KEY_STATE_RELEASED) {
-		if (key == KEY_SPACE) {
+	case WL_KEYBOARD_KEY_STATE_RELEASED:
+		if (keysym == XKB_KEY_space) {
 			state->edit_anchor = false;
 		}
+		break;
 	}
+
 }
 
 static const struct wl_keyboard_listener keyboard_listener = {
-	.keymap = noop,
+	.keymap = keyboard_handle_keymap,
```

Comment:
Also need to listen to the `modifiers` event.

Reply:
That's what I thought at first, but aren't modifiers just CTRL, Mod.... which means that we don't really care about them (for the moment), as the only keys we care about are space and escape

Reply:
The user may have configured his keyboard so that Shift+Caps Lock yields the Esc keysym, for instance.

Reply:
addressed